File: 2024/Day4/main.py
def count_xmas(grid):
    rows, cols = len(grid), len(grid[0])
    count = 0
    directions = [
        (0, 1), (1, 0), (1, 1), (-1, 1),
        (0, -1), (-1, 0), (-1, -1), (1, -1)
    ]

    def check_xmas(r, c, dr, dc):
        if (0 <= r < rows and 0 <= c < cols and
            0 <= r+3*dr < rows and 0 <= c+3*dc < cols):
            return (grid[r][c] == 'X' and
                    grid[r+dr][c+dc] == 'M' and
                    grid[r+2*dr][c+2*dc] == 'A' and
                    grid[r+3*dr][c+3*dc] == 'S')
        return False

    for r in range(rows):
        for c in range(cols):
            for dr, dc in directions:
                if check_xmas(r, c, dr, dc):
                    count += 1

    return count

def count_xmas2(grid):
    rows, cols = len(grid), len(grid[0])
    count = 0

    def check_xmas(r, c):
        if (r-1 < 0 or r+1 >= rows or c-1 < 0 or c+1 >= cols):
            return False
        center = grid[r][c]
        if center != 'A':
            return False
        corners = [
            grid[r-1][c-1], grid[r-1][c+1],
            grid[r+1][c-1], grid[r+1][c+1]
        ]
        return (corners.count('M') == 2 and
                corners.count('S') == 1 and
                ((corners[0] == 'M' and corners[3] == 'M') or
                 (corners[1] == 'M' and corners[2] == 'M')))

    for r in range(rows-1):
        for c in range(cols-1):
            if check_xmas(r, c):
                count += 1

    return count

with open('input.txt', 'r') as file:
    grid = [list(line.strip()) for line in file]

print(f"Part 1: {count_xmas(grid)}")
print(f"Part 2: {count_xmas2(grid)}")

File: 2024/Day4/main.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <vector>
int count_xmas(const std::vector<std::string>& grid) {
    int count = 0;
    int rows = grid.size();
    int cols = grid[0].size();

    auto check_xmas = [&](int r, int c, int dr, int dc) {
        if (r + 3*dr < 0 || r + 3*dr >= rows || c + 3*dc < 0 || c + 3*dc >= cols)
            return false;
        return grid[r][c] == 'X' && grid[r+dr][c+dc] == 'M' &&
               grid[r+2*dr][c+2*dc] == 'A' && grid[r+3*dr][c+3*dc] == 'S';
    };

    for (int r = 0; r < rows; ++r) {
        for (int c = 0; c < cols; ++c) {
            count += check_xmas(r, c, 0, 1);  // right
            count += check_xmas(r, c, 1, 0);  // down
            count += check_xmas(r, c, 1, 1);  // diagonal down-right
            count += check_xmas(r, c, 1, -1); // diagonal down-left
            count += check_xmas(r, c, 0, -1); // left
            count += check_xmas(r, c, -1, 0); // up
            count += check_xmas(r, c, -1, 1); // diagonal up-right
            count += check_xmas(r, c, -1, -1); // diagonal up-left
        }
    }

    return count;
}

int count_xmas2(const std::vector<std::string>& grid) {
    int count = 0;
    int rows = grid.size();
    int cols = grid[0].size();

    auto check_xmas = [&](int r, int c, int dr1, int dc1, int dr2, int dc2) {
        if (r + dr1 < 0 || r + dr1 >= rows || c + dc1 < 0 || c + dc1 >= cols ||
            r + dr2 < 0 || r + dr2 >= rows || c + dc2 < 0 || c + dc2 >= cols)
            return false;
        return (grid[r][c] == 'M' && grid[r+dr1][c+dc1] == 'A' && grid[r+dr2][c+dc2] == 'S') ||
               (grid[r][c] == 'S' && grid[r+dr1][c+dc1] == 'A' && grid[r+dr2][c+dc2] == 'M');
    };

    for (int r = 0; r < rows; ++r) {
        for (int c = 0; c < cols; ++c) {
            count += check_xmas(r, c, 1, -1, 2, -2) && check_xmas(r, c, 1, 1, 2, 2);
            count += check_xmas(r, c, 1, -1, 2, -2) && check_xmas(r+2, c-2, -1, 1, -2, 2);
            count += check_xmas(r, c, 1, 1, 2, 2) && check_xmas(r+2, c+2, -1, -1, -2, -2);
            count += check_xmas(r+2, c-2, -1, 1, -2, 2) && check_xmas(r+2, c+2, -1, -1, -2, -2);
        }
    }

    return count;
}

int main() {
    std::ifstream file("input.txt");
    std::vector<std::string> grid;
    std::string line;

    while (std::getline(file, line)) {
        grid.push_back(line);
    }

    int result = count_xmas(grid);
    int result2 = count_xmas2(grid);

    std::cout << "Part 1: " << result << std::endl;
    std::cout << "Part 2: " << result2 << std::endl;

    return 0;
}
